kubernetes.addons.dns: make corefile configurable

scvalex opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Add the `services.kubernetes.addons.dns.corefile` configuration option to set the corefile for the coredns addon in kubernetes.

Note that the difference between `default` and `defaultText` is that `${cfg.clusterDomain}` is interpolated in the former, but escaped in the latter.  If we didn't specify `defaultText` like this, the documentation would get regenerated on every build because `clusterDomain` might have changed.

###### Motivation for this change

In my case, I've found that coredns' caching is problematic, so I want to turn it off.  With this change, I can just do (note the commented out "cache" line):

```
      services.kubernetes.addons.dns.corefile = ".:10053 {
        errors
        health :10054
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :10055
        forward . /etc/resolv.conf
        # cache 30
        loop
        reload
        loadbalance
      }";
```

More generally, this lets you do more interesting things with coredns, like adding custom hosts, or tweaking the way hosts are resolved.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

I've run the `dns.nix` and `rbac.nix` kubernetes tests to make sure nothing is broken.  The tests still sometimes fail with "Error getting node", but that's nothing new.  I built the new documentation and checked that it looked good.

I'm also running this patch in my cluster and it does what it's supposed to.

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

scvalex commented:
@saschagrunert @johanot Could you review this PR, please?  It makes the coredns configuration customizable in the kubernetes addon.  I don't think it should have any effect on existing users.

scvalex commented:
@saschagrunert Sorry to a bother, but could you have a look at this?  Alternatively, do you know somebody else who might be a good reviewer?

saschagrunert reviewed (approved):
LGTM, sorry for the late reply

scvalex commented:
Thank you!

nh2 commented on nixos/modules/services/cluster/kubernetes/addons/dns.nix:
```diff
@@ -60,6 +60,45 @@ in {
+        .:${toString ports.dns} {
+          errors
+          health :${toString ports.health}
+          kubernetes \${cfg.clusterDomain} in-addr.arpa ip6.arpa {
```
I think it might be better to write:

```suggestion
          kubernetes \${config.services.kubernetes.addons.dns.clusterDomain} in-addr.arpa ip6.arpa {
```

so that the example shown in the manual can be literally pasted into your own config (currently `cfg` would not be in scope for that).

(For that it'd be handy if the `ports` were also a config options, so we could do the same there, but unfortunately they aren't, so we can't and changing that is out of the scope of this PR.)

scvalex replied:
Changed.

nh2 reviewed (commented):
This looks like a good idea to me; I added a question.

nh2 commented:
@ofborg test kubernetes

Following merge of #131694 (not clear if we have to rebase this on `master` for the tests to trigger, or if ofborg does the rebase for us automatically).

nh2 commented:
I'm not sure why on this PR I get

> warning: ignoring the user-specified setting 'restrict-eval', because it is a restricted setting and you are not a trusted user

usually I can start tests without problems.

scvalex committed:
kubernetes: make corefile configurable

scvalex commented on nixos/modules/services/cluster/kubernetes/addons/dns.nix:
```diff
@@ -151,20 +190,7 @@ in {
-            loop
-            reload
-            loadbalance
-          }";
```
It doesn't seem to make any difference and the tests pass in both cases.  I also tried `''` strings in my live deployment and that worked as well.

I changed this PR to use `''` since that seems to be the common choice for config files and it removes some unnecessary indentation in the resulting file.

scvalex commented:
@ofborg test kubernetes

scvalex commented:
@ofborg test kubernetes

Trying to run the tests again since they failed with the flaky "Error getting node": https://github.com/NixOS/nixpkgs/pull/128886/checks?check_run_id=3185563390

scvalex commented:
@ofborg test kubernetes.dns.singlenode kubernetes.dns.multinode kubernetes.rbac.singlenode kubernetes.rbac.multinode

scvalex commented:
To summarize where this PR is at the moment,
  - In addition to what @saschagrunert saw, I changed the corefile option to be a `''` string instead of a `"' string and expanded the path of the configuration option in the example.  I think these are non-contentious changes.
  - I changed the commit message to start with `kubernetes: ` instead of `kubernetes.addons.dns: `.  On the upside, the tests ran automatically.  On the downside, they failed with the usual flakyness, so I'm trying to figure out how to re-run them.  Calling `test kubernetes` doesn't work because the kubernetes tests are actually an attrset of an attrset of tests.  Calling `test kubernetes.dns.singlenode ...` works, but only for the first listed test.  I'm currently trying to figure out why that is.

scvalex commented:
@ofborg test kubernetes.dns.singlenode kubernetes.dns.multinode kubernetes.rbac.singlenode kubernetes.rbac.multinode

scvalex commented:
Third time's the charm, I guess.  The last rerun succeeded: https://github.com/NixOS/nixpkgs/pull/128886/checks?check_run_id=3186885692

That long command in my previous comment is the way to rerun the kubernetes tests, until we reorganize them a bit.  The only annoying bit is that it's so long that the github UI no longer shows the platform on this page, so it's hard to tell if you're looking at a failed x86_64 run, or a skipped aarch64 run.

Anyway, I think this features good to merge now, and the next step is to fix the tests.

nh2 commented:
>  The only annoying bit is that it's so long that the github UI no longer shows the platform on this page

I help myself with the hover toolip text:

![image](https://user-images.githubusercontent.com/399535/127492002-1f2dae05-ced3-43b0-b140-c57d40c9e502.png)
